tests/BookReader.test.js

```diff
@@ -1,59 +1,154 @@
+  test("Decodes a URI component and converts + to emptyStr", () => {
     expect(BookReader.util.decodeURIComponentPlus("https%3A%2F%2Farchive.org%2Fskr+")).toEqual("https://archive.org/skr ");
     expect(BookReader.util.decodeURIComponentPlus("%3Fx%3D+test")).toEqual("?x= test");
-});
+  });
 
-test('Encodes a URI component and converts emptyStr to +', () => {
+  test("Encodes a URI component and converts emptyStr to +", () => {
     expect(BookReader.util.encodeURIComponentPlus("?x=test ")).toEqual("%3Fx%3Dtest+");
     expect(BookReader.util.encodeURIComponentPlus("ABC abc 123")).toEqual("ABC+abc+123");
+  });
 });
 
-let clock;
+describe("Testing debounce function", () => {
+  let clock;
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers();
+  });
 
-beforeEach(() => {
-  clock = sinon.useFakeTimers();
+  afterEach(() => {
+    clock.restore();
+  });
+
+  test("testing debounce", () => {
+    const func = jest.fn();
+    const debouncedFunc = BookReader.util.debounce(func, 1000);
+    // Call it immediately
+    debouncedFunc();
+    expect(func).toHaveBeenCalledTimes(0); // func not called
+
+    // Call it several times with 500ms between each call
+    for (let i = 0; i < 10; i++) {
+      clock.tick(500);
+      debouncedFunc();
+    }
+    expect(func).toHaveBeenCalledTimes(0); // func not called
+
+    // wait 1000ms
+    clock.tick(1000);
+    expect(func).toHaveBeenCalledTimes(1); // func called
+  });
 });
 
-afterEach(() => {
-  clock.restore();
+describe('Testing Throttle function', () => {
+  let clock;
+
+  beforeEach(() => {
+    clock = sinon.useFakeTimers();
+  });
+  afterEach(() => {
+    clock.restore();
+  });
+
+  test('testing throttle', () => {
+    const callMethod = jest.fn();
+    const throttled = BookReader.util.throttle(() => callMethod(), 100);
+
+    const t = setInterval(() => {
+       throttled()
+    }, 50);
+
+    clock.tick(400);
+    expect(callMethod).toHaveBeenCalledTimes(4);
+    clearInterval(t);
+  });
 });
 
-test('testing debounce', () => {
-  const func = jest.fn();
-  const debouncedFunc = BookReader.util.debounce(func, 1000);
-  // Call it immediately
-  debouncedFunc();
-  expect(func).toHaveBeenCalledTimes(0); // func not called
+describe("Trigger function test", () => {
+  beforeEach(() => {
+    BookReader.prototype.trigger('SKR', 'val');
+  });
+  test("testing trigger", () => {
+    $(document).trigger('BookReader:' + 'SKR', 'val');
+  });
+});
 
-  // Call it several times with 500ms between each call
-  for(let i = 0; i < 10; i++) {
-    clock.tick(500);
-    debouncedFunc();
-  }
-  expect(func).toHaveBeenCalledTimes(0); // func not called
+describe("Testing extendParams", () => {
+  const obj1 = {
+    prop1: "SK",
+    prop2: 20
+  };
+  const obj2 = {
+    prop2: 19,
+    prop3: "SKR"
+  };
+  var modifiedNewParams = $.extend({}, obj2);
+  beforeEach(() => {
+    BookReader.prototype.extendParams(obj1, obj2);
```

Comment:
Instead of calling the `.prototype.extendParams()`, I would instantiate a BookReader object and call the method on it. ie:
```
var br = BookReader();
br.extendParams(obj1, obj2);
```
Also, it doesn't look like you're running and `expect` calls in this test. You probably want to create an expected dictionary and compare that to the result after calling `br.extendParams()`, ie:
```
var expectedResult = { foo: "bar", baz: "boop" };
var testObject = { foo: "bar" };
var extendedObject = { baz: "boop" };
br.extendParams(testObject, extendedObject);
expect(testObject).toEqual(expectedResult);
```

Reply:
Thank you for detailed review, this is very helpful